File: jp/ac/u_tokyo/t/eeic/seo/diagram/Edge.java
package jp.ac.u_tokyo.t.eeic.seo.diagram;

import java.util.Locale;

/**
 * 線分を定義.<br>
 * 線分とは異なる2点を通る直線のうち2点で挟まれる部分、ただし両端を含む
 * @author Seo-4d696b75
 * @version 2018/05/13
 */
public class Edge {

    public Edge(Point a, Point b){
        //ｘ(またはｙ)座標に関してa<=bを満たすように整理
        //ｘ座標が一致した場合はｙ座標に関して昇順
        if ( a.getX() < b.getX() ) {
            this.a = a;
            this.b = b;
        }else if ( a.getX() > b.getX()){
            this.a = b;
            this.b = a;
        }else{
            if ( a.getY() < b.getY() ){
                this.a = a;
                this.b = b;
            }else if ( a.getY() > b.getY() ) {
                this.a = b;
                this.b = a;
            }else{
                throw new IllegalArgumentException("point duplicated : " + a.toString());
            }
        }
    }

    public final Point a,b;

    private Line line;

    @Override
    public boolean equals(Object other){
        if ( other instanceof Edge ){
            Edge edge = (Edge)other;
            return a.equals(edge.a) && b.equals(edge.b);
        }
        return false;
    }

    @Override
    public int hashCode(){
        int result = 17;
        result = result * 31 + a.hashCode();
        result = result * 31 + b.hashCode();
        return result;
    }

    @Override
    public String toString(){
        return String.format(Locale.US, "Edge{%s-%s}", a.toString(), b.toString());
    }

    public Point getMiddlePoint(){
        return new BasePoint((a.getX() + b.getX())/2, (a.getY() + b.getY())/2);
    }

    public static boolean onEdge(Point start, Point end, Point p){
        return (start.getX() - p.getX()) * (end.getX() - p.getX()) + (start.getY() - p.getY()) * (end.getY() - p.getY()) <= 0 && Line.onLine(start, end, p);
    }

    /**
     * この線分を(index):(1-index)の比で内分する点を計算します.<br>
     * <strong>注意 </strong>{@link Point#compareTo(Point) 座標点の自然順序付け}によって決まるa<bの関係において、
     * 線分ABを(index):(1-index)の比で内分するので方向に注意する.{@code index<0}の場合は外分点になる.
     * @param index (0,1) => 内分点
     * @return Non Null
     */
    public Point getDivision(double index){
        return Point.getDivision(a, b, index);
    }

    public boolean onEdge(Point p){
        return onEdge(this.a, this.b, p);
    }

    /**
     * Equals to call {@link #toLine()} and {@link Line#onSameSide(Point, Point)}
     * @return
     */
    public boolean onSameSide(Point p1, Point p2){
        return toLine().onSameSide(p1, p2);
    }

    public Line toLine(){
        if ( line == null ){
            line = new Line(this);
        }
        return line;
    }

    /**
     * 線分との交点を取得する.
     * <strong>注意 </strong>交点を持たない場合はNull
     * @param line 交点を調べる直線
     * @return Null if no such intersection
     */
    public Point getIntersection(Line line){
        Point p = this.toLine().getIntersection(line);
        double v = (a.getX() - p.getX()) * (b.getX() - p.getX()) + (a.getY() - p.getY()) * (b.getY() - p.getY());
        if ( v <= 0 ){
            return p;
        }else{
            return null;
        }
    }

    /**
     * 線分との交点を取得する.
     * <strong>注意 </strong>交点を持たない場合はNull
     * @param edge 交点を調べる線分
     * @return Null if no such intersection
     */
    public Point getIntersection(Edge edge){
        Point p = edge.getIntersection(this.toLine());
        if ( p != null ){
            double v = (a.getX() - p.getX()) * (b.getX() - p.getX()) + (a.getY() - p.getY()) * (b.getY() - p.getY());
            if ( v <= 0 ){
                return p;
            }
        }
        return null;
    }

    public double getDistance(Point p){
        double v1 = (p.getX() - a.getX()) * (b.getX() - a.getX()) + (p.getY() - a.getY()) * (b.getY() - a.getY());
        double v2 = (p.getX() - b.getX()) * (a.getX() - b.getX()) + (p.getY() - b.getY()) * (a.getY() - b.getY());
        if ( v1 > 0 && v2 > 0 ){
            return toLine().getDistance(p);
        }else if ( v1 < 0 ){
            return p.measure(a);
        }else{
            return p.measure(b);
        }
    }

}

File: spherical/src/main/java/jp/seo/diagram/spherical/Edge.kt
package jp.seo.diagram.spherical

import java.util.*
import kotlin.math.acos

class Edge private constructor(
    val a: Vector,
    val b: Vector,
) {

    companion object {
        /**
         * ２点A,Bを結ぶ直線
         *
         * 直線の長さが測地的距離・大圏距離になる
         */
        fun between(a: Vector, b: Vector): Edge {
            if ((a - b).isZero) throw IllegalArgumentException("line can't be determined by duplicated points")
            if ((a + b).isZero) throw IllegalArgumentException("pair of antipodes can't determine a line")
            return if (a < b) Edge(a, b) else Edge(b, a)
        }

        fun between(a: LatLng, b: LatLng) = between(a.vec, b.vec)

        fun onEdge(a: Vector, b: Vector, p: Vector): Boolean {
            return Line.onLine(a, b, p)
                    && (acos(p.dot(a)) + acos(p.dot(b)) - acos(a.dot(b))).isZero
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        return other is Edge
                && a == other.a
                && b == other.b
    }

    override fun hashCode() = Objects.hash(a, b)

    override fun toString() = "Edge($a - $b)"

    val length: Double
        get() = acos(a.dot(b))

    val line: Line
        get() = Line.between(a, b)

    private fun onEdge(p: Vector): Boolean {
        val al = acos(a.dot(p))
        val bl = acos(b.dot(p))
        return (al + bl - length).isZero
    }

    fun intersection(other: Line): Vector? {
        val p1 = line.intersection(other)
        if (onEdge(p1)) return p1
        val p2 = -p1
        if (onEdge(p2)) return p2
        return null
    }

    fun intersection(other: Edge): Vector? {
        val p = other.intersection(line) ?: return null
        return if (onEdge(p)) p else null
    }
}